File: src/main/java/com/heekng/celloct/entity/Shop.java
package com.heekng.celloct.entity;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

import static javax.persistence.CascadeType.*;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Shop extends BaseTimeEntity {

    @Id @GeneratedValue
    @Column(name = "shop_id")
    private Long id;

    private String phone;
    private String name;
    private String info;

    @OneToMany(mappedBy = "shop", cascade = REMOVE)
    private List<Staff> staffList = new ArrayList<>();

    @OneToMany(mappedBy = "shop", cascade = REMOVE)
    private List<JoinRequest> joinRequests = new ArrayList<>();

    @OneToMany(mappedBy = "shop", cascade = ALL)
    private List<Manager> managers = new ArrayList<>();

    @Builder
    public Shop(String phone, String name, String info) {
        this.phone = phone;
        this.name = name;
        this.info = info;
    }

    public void update(String phone, String info) {
        this.phone = phone;
        this.info = info;
    }
}

File: src/main/kotlin/com/heekng/celloct/entity/Shop.kt
package com.heekng.celloct.entity

import javax.persistence.*

@Entity
class Shop(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY) @Column(name = "shop_id")
    val id: Long? = null,

    @Column(name = "phone")
    var phone: String? = null,

    @Column(name = "name", unique = true, nullable = false)
    val name: String,

    @Column(name = "info")
    var info: String? = null,

    @OneToMany(mappedBy = "shop", cascade = [CascadeType.REMOVE])
    val staffList: MutableList<Staff> = mutableListOf(),

    @OneToMany(mappedBy = "shop", cascade = [CascadeType.REMOVE])
    val joinRequests: MutableList<JoinRequest> = mutableListOf(),

    @OneToMany(mappedBy = "shop", cascade = [CascadeType.ALL])
    val managers: MutableList<Manager> = mutableListOf(),
) : BaseTimeEntity() {

    companion object {
        fun fixture(
            phone: String? = null,
            name: String,
            info: String? = null,
        ): Shop {
            return Shop(
                phone = phone,
                name = name,
                info = info
            )
        }
    }

    fun update(phone: String, info: String) {
        this.phone = phone
        this.info = info;
    }

    fun addManager(manager: Manager) {
        this.managers.add(manager)
    }
}
